design/src/rfcs/rfc0027_new_home_for_cred_cache_and_builder_api_consistency.md

```diff
@@ -0,0 +1,182 @@
+        provider
+    } else {
+        // --snip--
+    };
+
+    // We're still massaging the input parameters for `create_cache`
+    // and there could be more parameters, but what's important is
+    // that the method's return type is `SharedCredentialsCache`.
+    let credentials_cache: SharedCredentialsCache =
+        self.credentials_cache.create_cache(credentials_provider);
+
+    // --snip--
+
+    let mut builder = SdkConfig::builder()
+        .region(region)
+        .retry_config(retry_config)
+        .timeout_config(timeout_config)
+        .credentials_cache(credentials_cache) // new method
+        .http_connector(http_connector);
+
+    // --snip--
+
+    builder.build()
+}
+```
+
+The comment `// new method` above further implies that `aws_types::sdk_config::Builder` now has a method `credentials_cache` that takes a `SharedCredentialCache`. What this means is that `aws_types::sdk_config::Builder` has both `credentials_provider` and `credentials_cache` methods. This presents two problems.
+
+-  `aws_types::sdk_config::Builder` has to drop on the floor a credentials provider passed to the `credentials_provider` method.
+
+    There is another credentials provider already baked in a credentials cache passed from `aws_config::ConfigLoader::load`. `aws_types::sdk_config::Builder` has no way of creating another `SharedCredentialsCache` using the passed-in credentials provider to overwrite that credentials cache for the reason explained in the next problem.
+
+- Outside the `aws-config` crate, the user who directly interacts with `aws_types::sdk_config::Builder` has no way to prepare a `SharedCredentialCache` to pass it to the `credentials_cache` method.
+
+	The factory function for credentials caches, `CredentialsCache::create_cache`, is `pub(crate)` so the caches can only be created within the `aws-config` crate. Even if the factory function were marked as `pub`, the places that can produce the caches would still be limited to those that do not cause a cyclic crate dependency with respect to the `aws-config` crate, which excludes the `aws-types` crate for instance.
+
+Note that the same set of problems exists for a service client's `config::Builder`. Such a `Builder` has the methods `credentials_provider` and `set_credentials_provider`. Therefore, it is reasonable to expect that the `Builder` will also have the methods `credentials_cache` and `set_credentials_cache`, except that the assumption does not hold as the service client crate, being outside the `aws-config` crate, cannot create a credentials cache, unable to pass it to the said methods.
+
+To summarize, these are the problems encountered as a result of implementation spike for the selected design in the preceding RFC:
+- `ConfigLoader` has `credentials_provider` and `credentials_cache`, but that pair of methods does not seem to behave consistently in `aws_types::sdk_config::Builder` and every service client's `config::Builder`.
```

Comment:
I think I probably forgot to call out in the previous RFC that the builders should be identical all the way through, but that was my intent.

Reply:
Thanks for clarifying! As mentioned in the beginning, this RFC is meant to supplement the previous one 🙂 
Again, I can edit the previous RFC to add the disclaimer.

Reply:
Again, no need to edit the previous RFC. Only edited this RFC in 8706c43.